Persistence rework

svranesevic opened the pull request:
Persistence:
Replaced filesystem persistence with in-memory one
Enabled by default

model/Device:
Added 'const' to getters

README.md:
Added supported protocol(s)
Updated Persistence section

Service concept removed

Commit:
Persistence:
Replaced filesystem persistence with in-memory one
Enabled by default

model/Device:
Added 'const' to getters

README.md:
Added supported protocol(s)
Updated Persistence section

Service concept removed

Commit:
Wolk:
After publishing reschedule publish if there are leftover readings in persistence

Commit:
OutboundMessageFactory:
Removed exceptions

Commit:
README.md:
ActuationHandler, ActuatorStatusProvider explanation

Commit:
README.md:
Updated data publish strategy

Commit:
README.md:
Updated data publish strategy

milanzlicic commented on src/OutboundMessageFactory.cpp:
```diff
@@ -0,0 +1,172 @@
+
+    const json jPayload(sensorReadings);
+    const std::string payload = jPayload.dump();
+    const std::string topic = "readings/" + deviceKey + "/" + sensorReadings.front()->getReference();
```
Should define topic strings somewhere and use them here

milanzlicic commented on src/OutboundMessageFactory.cpp:
```diff
@@ -0,0 +1,172 @@
+    const std::string payload = jPayload.dump();
+    const std::string topic = "actuators/status/" + deviceKey + "/" + actuatorStatuses.front()->getReference();
+
+    return std::make_shared<OutboundMessage>(payload, topic, 1);
```
What does 1 represent?

milanzlicic commented on src/Wolk.h:
```diff
@@ -95,17 +97,32 @@ class Wolk
     static unsigned long long int currentRtc();
 
+    void publishActuatorStatuses();
+    void publisAlarms();
```
missing a 'h'

milanzlicic commented on src/persistence/inmemory/InMemoryPersistence.cpp:
```diff
@@ -0,0 +1,168 @@
+
+bool InMemoryPersistence::putAlarm(const std::string& key, std::shared_ptr<Alarm> alarm)
+{
+    getOrCreateAlarmsByKey(key).push_back(alarm);
```
Missing return statement

milanzlicic commented on src/CommandBuffer.h:
```diff
@@ -0,0 +1,102 @@
+
+namespace wolkabout
+{
+template <class Command> class CommandBuffer
```
Don't make it template, just let the command be std::function<void()>

milanzlicic commented on src/Wolk.cpp:
```diff
@@ -87,56 +91,143 @@ void Wolk::addAlarm(const std::string& reference, const std::string& value, unsi
-, m_publishService(publishService)
 {
+    m_commandBuffer = std::unique_ptr<CommandBuffer<std::function<void()>>>(new CommandBuffer<std::function<void()>>());
+    m_commandBufferThread = std::thread([this]() -> void {
```
Should stop and join thread in destructor

Commit:
CommandBuffer:
No longer template, Command is of type 'std::function<void()>'
Join on m_worker thread in destructor to avoid calling of std::terminate() (See http://en.cppreference.com/w/cpp/thread/thread/~thread)

Commit:
Wolk:
publishAlarms typo fix

Commit:
InMemoryPersistence:
Added missing return statement

Commit:
OutboundMessageFactory:
Extracted topic roots to constants
Explained 'magic' number in 'make' from actuator statuses

MqttConnectivityService:
Removed unused constants

Commit:
CommandBuffer:
Formatting

OutboundMessageFactory:
Formatting
